### ISSUE TITLE
Fix S2583 FP: Counter increment and condition always false on loop unrolling

### ISSUE THREAD

**Issue by bart-vmware** [SonarSource/sonar-dotnet#8139]
### Description

We've found cases in our codebase where Sonar incorrectly suggests replacing a condition with `true`, claiming that the code is unreachable, which is incorrect.

### Repro steps

1. `dotnet new console -f net6.0`
2. `dotnet add package SonarAnalyzer.CSharp`
3. Paste the code below in `Program.cs`:
  ```c#
  int counter = 0;
  
  for (int index = 0; index < 8; index++)
  {
      if (counter < 4) // S2583: Conditionally executed code should be reachable
      {
          Console.WriteLine($"TRUE: counter={Get(counter++)}");
      }
      else
      {
          Console.WriteLine($"FALSE: counter={Get(counter++)}");
      }
  }
  
  static int Get(int value)
  {
      return value;
  }
  ```

Running this program prints:
```
TRUE: counter=0
TRUE: counter=1
TRUE: counter=2
TRUE: counter=3
FALSE: counter=4
FALSE: counter=5
FALSE: counter=6
FALSE: counter=7
```

Changing the `if` line to `if (true)`, as suggested by the rule description, changes the output to:
```
TRUE: counter=0
TRUE: counter=1
TRUE: counter=2
TRUE: counter=3
TRUE: counter=4
TRUE: counter=5
TRUE: counter=6
TRUE: counter=7
```

Which shows different behavior.

### Expected behavior

No warning, because the code is reachable.

### Actual behavior

Warning that suggests breaking the program behavior.

### Related information

* C#/VB.NET Plugins version: N/A
* Visual Studio version: v17.4.4
* MSBuild / dotnet version: .NET 6, SDK v7.0.401
* SonarScanner for .NET version: v9.11
* Operating System: Windows

**Comment by antonioaversa** [comment]
Hi @bart-vmware, thanks for reporting the issue.

I confirm I can reproduce the False Positive. The rule [S2583](https://sonarsource.github.io/rspec/#/rspec/S2583/csharp) is based on symbolic execution: it analyses the flow of operations and determines solution spaces for the variables in the state of the program.

In the presence of loops, we run the flow for a limited number of iterations. In your scenario, the condition `counter < 4` evaluates to `true` at every symbolic execution stage, and that's why we wrongly conclude that the condition always evaluates to `true`, rendering the `else` unreachable.

Similar issues with loop unrolling have been reported recently on this rule, and we are planning a dedicated sprint later this year to tackle all such issues in a consistent way. I am adding this issue to our backlog, to be included in the planning of the sprint.

Apologies for the inconvenience. For the time being, waiting for the issue to be fixed:
- If you are using SonarLint with SonarQube or SonarCloud in connected mode: you can mark the issue as a false-positive in the UI. The rule can also be disabled in your Quality Profile if it’s particularly noisy.
- If you are using SonarLint in Standalone Mode: you can use `#pragma warning disable` and `#pragma warning restore` to locally disable the rule in the source, or the `.editorconfig` files to disable or change the severity of the rule, if it gets particularly noisy.

**Comment by Tim-Pohlmann** [comment]
I close this as a duplicate of #8028 to track all issues caused by loop counters not being fully explored in one place.

**Comment by bart-vmware** [comment]
Please prioritize fixing this, it is a very dangerous bug.

**Comment by bart-vmware** [comment]
Thanks
